File: prototype/rails/src/test/actions/listings/change_listing_status_test.rb
require "test_helper"

module Listings
  class ChangeListingStatusTest < ActiveSupport::TestCase
    test "a draft goes on sale" do
      art = create_listing(status: Domain::Listings::ListingStatus::DRAFT)

      ChangeListingStatus.new.call(listing: art, status: Domain::Listings::ListingStatus::FOR_SALE)

      assert_equal Domain::Listings::ListingStatus::FOR_SALE, art.reload.status
    end

    test "a listing on sale is archived" do
      art = create_listing(status: Domain::Listings::ListingStatus::FOR_SALE)

      ChangeListingStatus.new.call(listing: art, status: Domain::Listings::ListingStatus::ARCHIVED)

      assert_equal Domain::Listings::ListingStatus::ARCHIVED, art.reload.status
    end

    test "a move the lifecycle refuses raises" do
      art = create_listing(status: Domain::Listings::ListingStatus::DRAFT)

      assert_raises(Domain::TransitionError) do
        ChangeListingStatus.new.call(listing: art, status: Domain::Listings::ListingStatus::SOLD)
      end

      assert_equal Domain::Listings::ListingStatus::DRAFT, art.reload.status
    end
  end
end

File: prototype/php/src/app/Actions/Listings/ChangeListingStatusTest.php
<?php

declare(strict_types=1);

namespace App\Actions\Listings;

use App\Domain\Listings\ListingStatus;
use DomainException;

it('moves a listing through an allowed transition', function (ListingStatus $from, ListingStatus $to): void {
    $listing = $this->listing($this->seller(), ['status' => $from]);

    app(ChangeListingStatus::class)($listing, $to);

    expect($listing)->toHaveStatus($to);
})->with([
    'draft to for sale' => [ListingStatus::Draft, ListingStatus::ForSale],
    'for sale to archived' => [ListingStatus::ForSale, ListingStatus::Archived],
]);

it('refuses a transition the lifecycle does not allow', function (): void {
    $listing = $this->listing($this->seller(), ['status' => ListingStatus::Archived]);

    app(ChangeListingStatus::class)($listing, ListingStatus::ForSale);
})->throws(DomainException::class);

it('leaves the row alone when the transition is refused', function (): void {
    $listing = $this->listing($this->seller(), ['status' => ListingStatus::Draft]);

    try {
        app(ChangeListingStatus::class)($listing, ListingStatus::Sold);
    } catch (DomainException) {
        // The assertion below is the point; the throw is asserted above.
    }

    expect($listing)->toHaveStatus(ListingStatus::Draft);
});
